#!/usr/bin/env ruby

require 'sandbox'
require 'test/unit'

class TestExploits < Test::Unit::TestCase
  def setup
    @s = Sandbox.safe
  end

  def test_safe_return
    k = @s.eval("Kernel")
    def k.crack
      open("/etc/passwd")
    end
    assert Kernel.respond_to?(:crack)
    assert ! @s.eval("Kernel.respond_to?(:crack)")
  end

  def test_safe_eval
    assert ! @s.eval("eval('Kernel').respond_to?(:fork)")
    assert ! @s.eval("eval('Object.const_defined? :TOPLEVEL_BINDING')")
  end

  def test_safe_globals
    assert_equal "(sandbox)", @s.eval("$0")
    /(.)(.)(.)/.match("abc")
    assert_equal "TEST", @s.eval("$TEST = 'TEST'; $TEST")
    assert_equal "e", @s.eval("/(.)(.)(.)/.match('def'); $2")
    assert_equal "b", $2
    assert_equal "TEST", @s.eval("$TEST")
    assert_equal "e", @s.eval("$2")
    /(.)(.)(.)/.match("ghi")
  end

  def test_fork
    assert_raise Sandbox::Exception do
      @s.eval("Kernel.fork")
    end
    assert_equal :fork, @s.eval("begin; Kernel.fork; rescue NameError; :fork; end"),
      "DANGEROUS EXPLOIT: Kernel#fork available in the sandbox."
    assert_equal :fork, @s.eval("begin; fork; rescue NameError; :fork; end"),
      "DANGEROUS EXPLOIT: Kernel#fork available in the sandbox."
  end

  def test_io
    assert_equal :NameError, @s.eval("begin; File.open('/etc/passwd'); rescue NameError; :NameError; end"),
      "DANGEROUS EXPLOIT: Filesystem can be accessed!"
  end

  def test_globals
    assert !@s.eval("$0.class.ancestors.last.respond_to?(:p)"),
      "DANGEROUS EXPLOIT: Kernel can be reached through the globals."
  end
  
  def test_block_scope
    1.times do |i|
      assert [], @s.eval("local_variables")
    end
  end

  def test_matchdata_and_friends
    assert_equal :NameError, 
      @s.eval("begin; /(.+)/.match('FreakyFreaky').instance_eval { open('/etc/passwd') }; rescue NameError; :NameError; end"),
        "DANGEROUS EXPLOIT: MatchData references Kernel."
    assert_equal :NameError, 
      @s.eval("begin;(begin;Regexp.new('(');rescue e;e;end).instance_eval{ open('/etc/passwd') }; rescue NameError; :NameError; end"),
        "DANGEROUS EXPLOIT: RegexpError references Kernel."
  end

  # XXX: THIS ONE'S A HUGE PROBLEM
  # def test_stack_level
  #   assert ! @s.eval("begin; p = proc { |x| p[x] }; p[1]; ensure; global_variables.include?('$!'); end")
  # end
  # def test_caller
  #   assert ! @s.eval("caller[0].class.ancestors.last.respond_to?(:p)")
  # end
end
